### PR TITLE
Feature/provision local service

### PR DESCRIPTION
**Pull request by jwwwb**
## Describe changes
I implemented/fixed _ to achieve _.

## Pre-requisites
Please ensure you have done the following:
- [ ] I have read the **CONTRIBUTING.md** document.
- [ ] If my change requires a change to docs, I have updated the documentation accordingly.
- [ ] If I have added an integration, I have updated the [integrations](https://docs.zenml.io/features/integrations) table.
- [ ] I have added tests to cover my changes.

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Other (add details above)

### CODE REVIEW

File: src/zenml/integrations/mlflow/services/mlflow_deployment.py [hunk 1]
```diff
@@ -185,9 +185,12 @@ def predict(self, request: "NDArray[Any]") -> "NDArray[Any]":
...
-            self.endpoint.prediction_uri,
-            json={"instances": request.tolist()},
-        )
+        if self.endpoint.prediction_uri is not None:
+            response = requests.post(
+                self.endpoint.prediction_uri,
+                json={"instances": request.tolist()},
+            )
+        else:
+            raise ValueError("No endpoint known for prediction.")
```
Comment (AlexejPenner) [inline src/zenml/integrations/mlflow/services/mlflow_deployment.py:194]: How are these changes related to our service? Was mypy complaining?

Comment (jwwwb) [reply src/zenml/integrations/mlflow/services/mlflow_deployment.py:194]: Indeed it was. `self.endpoint` is an `Optional`, so the `.prediction_uri` does not always exist

File: src/zenml/stack_stores/rest_stack_store.py [hunk 2]
```diff
@@ -0,0 +1,161 @@
...
+        """Fetches a stack."""
+        return StackWrapper.parse_obj(self.get(f"/stacks/{name}"))
+
+    def register_stack(self, stack: StackWrapper) -> Dict[str, str]:
+        """Registers a stack and its components."""
+        return cast(Dict[str, str], self.post("/stacks/register", stack))
+
+    def deregister_stack(self, name: str) -> None:
+        """Deregisters a stack."""
+        self.get(f"stacks/{name}/deregister")
```
Comment (schustmi) [inline src/zenml/stack_stores/rest_stack_store.py:99]: Would a delete request make more sense here?

Comment (schustmi) [inline src/zenml/stack_stores/rest_stack_store.py:99]: Would a delete request make more sense here?

File: src/zenml/stack_stores/rest_stack_store.py [hunk 3]
```diff
@@ -0,0 +1,161 @@
...
+        self.get(f"/components/deregister/{component_type}/{name}")
+
+    # don't call these! ABC complains that they aren't implemented, but they
+    # aren't needed with custom implementations of the above:
+
+    def _create_stack(
+        self, name: str, stack_configuration: StackConfiguration
+    ) -> None:
+        """Add a stack to storage"""
+        raise NotImplementedError("Not to be accessed direclty in client!")
```
Comment (schustmi) [inline src/zenml/stack_stores/rest_stack_store.py:131]: `directly` type here and in the following methods

Comment (schustmi) [inline src/zenml/stack_stores/rest_stack_store.py:131]: `directly` type here and in the following methods

File: src/zenml/zen_service/zen_service_deployment.py [hunk 4]
```diff
@@ -0,0 +1,139 @@
...
+        except KeyboardInterrupt:
+            logger.info("Zen service stopped. Resuming normal execution.")
+
+    @property
+    def zen_service_uri(self) -> Optional[str]:
+        """Get the URI where the service is running.
+
+        Returns:
+            The URI where the service can be contacted
+            HTTP/REST inference requests, or None, if the service isn't running.
```
Comment (schustmi) [inline src/zenml/zen_service/zen_service_deployment.py:135]: Incoherent sentence, probably from some mlflow copy-paste?

Comment (schustmi) [inline src/zenml/zen_service/zen_service_deployment.py:135]: Incoherent sentence, probably from some mlflow copy-paste?

File: src/zenml/stack_stores/rest_stack_store.py [hunk 5]
```diff
@@ -0,0 +1,161 @@
...
+            for c in self.get(f"/components/{component_type}")
+        ]
+
+    def deregister_stack_component(
+        self, component_type: StackComponentType, name: str
+    ) -> None:
+        """Deregisters a stack component."""
+        self.get(f"/components/deregister/{component_type}/{name}")
+
+    # don't call these! ABC complains that they aren't implemented, but they
```
Comment (schustmi) [inline src/zenml/stack_stores/rest_stack_store.py:124]: So all the private methods below aren't needed as this interface overwrites the public-facing ones of the base class with it's own logic? Will this be beneficial for us in the long-term or will we frequently run into problems (e.g someone implements a new public method on the `BaseStackStore` which calls one of the unimplemented ones, we'd have to somehow catch that it needs to be implemented in the RestStackStore as well). Or coming from another angle, what's the downside to implementing the private methods instead? More HTTP requests?

Comment (schustmi) [inline src/zenml/stack_stores/rest_stack_store.py:124]: So all the private methods below aren't needed as this interface overwrites the public-facing ones of the base class with it's own logic? Will this be beneficial for us in the long-term or will we frequently run into problems (e.g someone implements a new public method on the `BaseStackStore` which calls one of the unimplemented ones, we'd have to somehow catch that it needs to be implemented in the RestStackStore as well). Or coming from another angle, what's the downside to implementing the private methods instead? More HTTP requests?

File: src/zenml/zen_service/zen_service.py [hunk 6]
```diff
@@ -0,0 +1,122 @@
...
+
+app = FastAPI()
+
+# to run this, execute:
+# uvicorn zenml.zen_service.zen_service:app --reload
+
+
+root: Path = Path(get_global_config_directory())
+url = f"sqlite:///{root / 'service_stack_store.db'}"
+print(url)
```
Comment (schustmi) [inline src/zenml/zen_service/zen_service.py:25]: Maybe some explanation what the URL means? I think also some docstrings in this file would probably be nice to show up in the API-docs, even if it's just a one-liner each

Comment (schustmi) [inline src/zenml/zen_service/zen_service.py:25]: Maybe some explanation what the URL means? I think also some docstrings in this file would probably be nice to show up in the API-docs, even if it's just a one-liner each

**Review by schustmi — CHANGES_REQUESTED** [review]

File: src/zenml/zen_service/zen_service.py (hunk 6 above)
Comment (jwwwb) [reply src/zenml/zen_service/zen_service.py:25]: Yeah a lot of this is still incomplete / WIP just pushed for the sake of collaboration, let's see how we can make this land in the API docs.

File: src/zenml/zen_service/zen_service_deployment.py [hunk 7]
```diff
@@ -0,0 +1,139 @@
...
+)
+
+logger = get_logger(__name__)
+
+ZEN_SERVICE_URL_PATH = "/"
+ZEN_SERVICE_HEALTHCHECK_URL_PATH = "health"
+
+
+class ZenServiceEndpointConfig(LocalDaemonServiceEndpointConfig):
+    """MLflow daemon service endpoint configuration.
```
Comment (AlexejPenner) [inline src/zenml/zen_service/zen_service_deployment.py:24]:
```suggestion
    """Zenml daemon service endpoint configuration.
```

File: src/zenml/zen_service/zen_service_deployment.py [hunk 8]
```diff
@@ -0,0 +1,139 @@
...
+
+    Attributes:
+        zen_service_uri_path: URI path for the zenml service
+    """
+
+    zen_service_uri_path: str
+
+
+class ZenServiceEndpoint(LocalDaemonServiceEndpoint):
+    """A service endpoint exposed by the MLflow deployment daemon.
```
Comment (AlexejPenner) [inline src/zenml/zen_service/zen_service_deployment.py:34]:
```suggestion
    """A service endpoint exposed by the Zenml Service deployment daemon.
```

File: src/zenml/zen_service/zen_service_deployment.py (hunk 4 above)
Comment (AlexejPenner) [inline src/zenml/zen_service/zen_service_deployment.py:135]:
```suggestion
            The URI where the service can be contacted via
            HTTP/REST inference requests, or None, if the service isn't running.
```

**Review by AlexejPenner — CHANGES_REQUESTED** [review]
Caught some copy-pasta mistakes
